Various extensions and improvements for asserts

bernt-matthias opened the pull request:
### Changes to text, size, tabular, archive, and xml assertions: 

- Implement `n`, `delta`, `min`, `max` for all of them (previously only `delta` in `has_size`). I hope that we can save lots of test data files, since this can replace `<output compare="sim_size">`.
  - for xml only the new generalizing `xml_element` assertion and `has_n_elements_with_path` support `n`, `delta`, `min`, `max`
  - added new Bytes type to xsd to allow for bytes suffixes for all these nubers (might be useful also at other places)
- Implement `negate` attribute

### Changes to `has_n_columns`

wanted to use a `has_n_columns` for this file https://github.com/galaxyproject/tools-iuc/blob/master/tools/stacks2/test-data/populations/populations.sumstats.tsv which is not useful due to the leading comment lines which have different numbers of columns. This PR adds a new attribute `comment` to the tag which allows to ignore comment lines for the assertion. 

Furthermore: 

- only non-empty lines are considered (this can be changed)

### xml assertions

- add `xml_element` assertion which generalizes all other xml assertions
- all xml assertions are now implemented by a call to `assert_xml_element`

### Tests and xsd changes

- adds unit tests for all  assertions
- correctly represent assertions in xsd which allows for better linting
  - also show schema doc for assertions not as table but analogous to what is done for param (more readable), each with a table of allowed attributes
- annotate types (int and float) for assertion functions
- add linting of asserts

### More

- fix a bug in `element_text_is` and `attribute_is`: they also allowed only a prefix matching

TODO (input needed): 

- [ ] create a pr with the assert tests from dev to check if functionality has changed (only those that already worked, checking for `len(x) == 1` to allow for reformulation)
- [ ] command, stdout and stderr assertions should maybe only allow text (possibly also tabular) assertions?
  - would break backward compatibility
- [x] assertions with sub-assertions make no sense if negate=True
  - [x] should be covered by linter
- [x] would actually be cool to allow `n`, `delta`, `min`, `max` to allow for `500k`, `50M`, ...
- [ ] the introduced type annotations in the assertion functions are kind of odd, since the functions still get `str` (because the parser does not parse them to `int`,...), 
  - but they are helpful for the linter which now uses the annotated types
  - which brings me to the question if the linter needs to check the types because they are now also annotated in xsd: -> **xsd covers this .. is this enough?**
- [ ] some of the text assertions check for `None` (i.e. this is currently not consistent) .. but I have no idea what it is good for https://github.com/galaxyproject/galaxy/blob/7335edd0e853e4d09b747f101b2f5520de9c34fb/lib/galaxy/tool_util/verify/asserts/text.py#L8 .. we should use it in all or none of the assertions.
  - [ ] does it make sense to have this assert not affected by `negate`?
  - [ ] should be reflected in the xsd docs
  - [ ] This is most likely because the `has_archive_member` and `element_text` passed `None` to contained assertions. This has been changed .. so not necessary  anymore
- [ ] add delta and min / max to all assertions having `n`/`value`
  - [x] text, column, general
  - [x] archive
  - [x] xml
    - only for `xml_element` assertion and `has_n_elements_with_path` 
- [ ] implement `negate`
  - [x] text, column, general
  - [x] archive
  - [x] xml

## How to test the changes?

(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

bernt-matthias committed:
make has_n_columns check only non-empty and non-comment lines

wanted to use a has_n_columns assertion for this file:

https://github.com/galaxyproject/tools-iuc/blob/master/tools/stacks2/test-data/populations/populations.sumstats.tsv

this file might be problematic since Galaxy's tabular data type
uses the first comment line for column definitions

Also makes the assert text similar to the has_n_lines text

bernt-matthias committed:
fix linter problem

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/tabular.py:
```diff
@@ -16,4 +19,4 @@ def assert_has_n_columns(output, n, sep='\t'):
     first_line = get_first_line(output)
     assert first_line is not None, "Was expecting output with %d columns, but output was empty." % n
-    assert len(first_line.split(sep)) == n, "Output does not have %d columns." % n
+    assert len(first_line.split(sep)) == n, f"Expected {n} columns in output, found {first_line.split(sep)} lines"
```
```suggestion
    n_columns = len(first_line.split(sep))
    assert n_columns == n, f"Expected {n} columns in output, found {n_columns} columns"
```

Can we add a parameter here that controls whether we skip non-comment and non-empty lines (and what the comment character is) ? Alternatively, can we get the number of columns form the dataset metadata ?

bernt-matthias replied:
> Alternatively, can we get the number of columns form the dataset metadata ?

This would be nice, but I think the functions get plain text file content.

> Can we add a parameter here that controls whether we skip non-comment and non-empty lines (and what the comment character is) ?

How about adding a `comment` attribute which is empty by default. If given lines starting with any of the characters are ignored.

bernt-matthias committed:
fix assertion text

bernt-matthias committed:
add unit tests for tabular assertions

bernt-matthias committed:
document sep attribute of has_n_columns

bernt-matthias committed:
fixes

bernt-matthias committed:
allow to specify comment char for has_n_columns

bernt-matthias committed:
add AssertHasNColumns type in xsd

bernt-matthias committed:
add unit tests for text asserts and proper xsd

bernt-matthias committed:
add delta for has_n_lines and has_size

and delta_frac for has_size

plus tests and xsd changes

bernt-matthias committed:
fix xsd

bernt-matthias committed:
add delta_frac also for has_n_lines

and

- reformulations in xsd file
- consistent use of `.` in assertion messages

bernt-matthias committed:
document sim_size as discouraged

bernt-matthias committed:
simplify assertion code a bit

do not store modules, but directly the functions

bernt-matthias committed:
annotate int and float params of assertion functions

bernt-matthias committed:
add linting of assertions

bernt-matthias committed:
remove print and fix unit test

bernt-matthias committed:
fix numbers in assertion test

bernt-matthias committed:
add n for assert_has_text_matching and assert_has_line_matching

bernt-matthias committed:
add delta min max to text, tabular, and size assertions

for tabular and size n/value becomes optional

and remove delta_frac

bernt-matthias committed:
xml assertions: tests, small improvements + xsd reformatting

- added unit tests for all xml assertions
- restructured xsd:
  - grouped general, text, tabular, archive, xml, and h5 assertions
  - full definition of all assertions in xsd
  - allow for "recursion" in xsd for has_archive_member and element_text
- restructured schema doc: instead of the table its now a
  definition list (no more horizontal scrolling in the wide table)
  also here assertions are grouped

bernt-matthias committed:
add negate attribute to text, size, tabular assertions

- assertions messages have been reformulated to make them negatable
  (which also fixes the unit tests for the previous commit
   which did parallel edits to the test)

bernt-matthias committed:
xsd: use AttributeGroup for some assertion attributes

to make the xsd more compact and less redundant

bernt-matthias committed:
add tests for hdf5 and archive asserts and some fixes

archive asserts:

- check for absent members directly in the assertion
  - before the _extract... functions returned `None`
    which had no effect if there is no included assertion
    (probably this is also the reason why some of the
     text assertions check for None)
  - so now the assertion can be used without included
    sub-assertions to check for member presence
- fix for non-file members: the behavior for zip and tar
  was different

hdf5 asserts:

- fix assertion text

bernt-matthias committed:
xml assert: element_text check if path exists

bernt-matthias committed:
reformulate has_archive_member docs

bernt-matthias committed:
conditionally execute h5 assertion tetst

if h5py module is available. otherwise package tests are failing

bernt-matthias committed:
test archive assert using no test data

but just create mock files on the fly

bernt-matthias committed:
add all attribute to archive asserts

allowing to check sub assertions for all archive members matching path

bernt-matthias committed:
fix/complete archive test for sub assertions

bernt-matthias committed:
followup on xsd restructuring

doc parsing: asserts are now choice

bernt-matthias committed:
add n, delta, min, max for archive assertions

bernt-matthias committed:
fix element_text_matches and attribute_matches

the assertions also accepted that the given text is a prefix of
the attribute/text

bernt-matthias committed:
introduce xml_element assertion

which generalizes all other xml assertions

bernt-matthias committed:
more specific error message for archive assertions

and

- has_n_elements_with_path add delta, min, max
- negate for all xml assertions (except is_valid)

bernt-matthias committed:
schema doc: add attribute table to all assertions

bernt-matthias committed:
xsd fix

bernt-matthias committed:
allow byte suffixed for numbers in asserts

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/__init__.py:
```diff
@@ -12,16 +12,19 @@
+assertion_functions = {}
 for assertion_module_name in assertion_module_names:
-    full_assertion_module_name = f"galaxy.tool_util.verify.asserts.{assertion_module_name}"
+    full_assertion_module_name = 'galaxy.tool_util.verify.asserts.' + assertion_module_name
```
I did prefer the f-string, it's more readable and theoretically slightly faster

bernt-matthias replied:
```suggestion
    full_assertion_module_name = f"galaxy.tool_util.verify.asserts.{assertion_module_name}"
```

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/__init__.py:
```diff
@@ -12,16 +12,19 @@
 # source file "test/base/asserts/<MODULE_NAME>.py" and add
 # <MODULE_NAME> to the list of assertion module names defined above.
-assertion_modules = []
+assertion_functions = {}
```
@simonbray that's going to affect https://github.com/galaxyproject/planemo/pull/1213/, I guess. This should be easier to use down the line though.

mvdbeek commented on lib/galaxy/tool_util/verify/asserts/__init__.py:
```diff
@@ -33,14 +36,11 @@ def verify_assertions(data, assertion_description_list):
-        if hasattr(assertion_module, assert_function_name):
-            assert_function = getattr(assertion_module, assert_function_name)
+    assert_function_name = "assert_" + tag
+    assert_function = assertion_functions.get(assert_function_name, None)
```
```suggestion
    assert_function = assertion_functions.get(assert_function_name)
```

bernt-matthias committed:
Update lib/galaxy/tool_util/verify/asserts/__init__.py

Co-authored-by: Marius van den Beek <m.vandenbeek@gmail.com>

bernt-matthias committed:
Update lib/galaxy/tool_util/verify/asserts/__init__.py

bernt-matthias committed:
fix rebase errors

mvdbeek reviewed (approved):
Do you think this is ready to merge @bernt-matthias ?

jmchilton commented on test/unit/tool_util/verify/test_asserts.py:
```diff
@@ -0,0 +1,1012 @@
+import os
+import shutil
```
Amazing!
